lib/utils/display.js

```diff
@@ -344,22 +366,118 @@ class Display {
         this.#logColors[level](level),
         title ? this.#logColors.title(title) : null,
       ]
-      this.#stderrWrite({ prefix }, ...args)
-    } else if (this.#progress) {
-      // TODO: make this display a single log line of filtered messages
+      this.#write(this.#stderr, { prefix }, ...args)
+    }
+  }
+}
+
+class Progress {
+  // Taken from https://github.com/sindresorhus/cli-spinners
+  // MIT License
+  // Copyright (c) Sindre Sorhus <sindresorhus@gmail.com> (https://sindresorhus.com)
+  static dots = { duration: 80, frames: ['⠋', '⠙', '⠹', '⠸', '⠼', '⠴', '⠦', '⠧', '⠇', '⠏'] }
+  static lines = { duration: 130, frames: ['-', '\\', '|', '/'] }
+
+  #stream
+  #spinner
+  #enabled = false
+
+  #frameIndex = 0
+  #lastUpdate = 0
+  #interval
+  #timeout
+
+  // We are rendering is enabled option is set and we are not waiting for the render timeout
+  get #rendering () {
+    return this.#enabled && !this.#timeout
+  }
+
+  // We are spinning if enabled option is set and the render interval has been set
+  get #spinning () {
+    return this.#enabled && this.#interval
+  }
+
+  constructor ({ stream }) {
+    this.#stream = stream
+  }
+
+  load ({ enabled, unicode }) {
+    this.#enabled = enabled
+    this.#spinner = unicode ? Progress.dots : Progress.lines
+    // Dont render the spinner for short durations
+    this.#render(200)
```

Comment:
What's the performance cost here?  Is 500 still readable but measurably faster?  Ultimately if performance is that important folks should turn this off but ... a sensible default should be something that comes from a place of some testing.

Reply:
I don't think it should affect performance. This delay is more for UI. A command that only takes 100ms shouldn't attempt to render the spinner since it won't appear as more than a flash of a spinner. I landed on 200ms as the "if nothing has been output yet, let the user know that things are happening". But this number should be very open to changing in the future.

Reply:
Reading the next comment, I see what you're talking about with the interval frame rate. That number is located in the `Progress.dots.duration` and `Progress.lines.duration` and we should also feel free to experiment with those as well.

Reply:
9da57388ebd5c643c2a95bbf63abc745cad45ccc

Reply:
d9cf7fcf16950c1bb885cdc99d3ba9e9c400e7e1